scanner/scanner.go

```diff
@@ -108,6 +117,100 @@ func (s *Scanner) ScanAndClean(opts ScanOptions) (*Context, error) {
 	return c, nil
 }
 
+func (s *Scanner) ExecuteWatch() error {
+	var err error
+	s.watcher, err = fsnotify.NewWatcher()
+	if err != nil {
+		log.Fatalf("error creating watcher: %v\n", err)
+	}
+	defer s.watcher.Close()
+
+	for _, dir := range s.musicDirs {
+		err := filepath.WalkDir(dir, func(absPath string, d fs.DirEntry, err error) error {
+			return s.watchCallback(dir, absPath, d, err)
+		})
+		if err != nil {
+			log.Printf("error watching directory tree: %v\n", err)
+		}
+	}
+
+	if _, err = s.ScanAndClean(ScanOptions{}); err != nil {
+		log.Printf("error scanning: %v", err)
+	}
+
+	for {
+		select {
+		case event := <-s.watcher.Events:
+			doScan := event.Op&(fsnotify.Create|fsnotify.Write) != 0
+			if doScan && s.StartScanning() {
+				var dirName string
+				c := &Context{
+					errs:       &multierr.Err{},
+					seenTracks: map[int]struct{}{},
+					seenAlbums: map[int]struct{}{},
+					isFull:     false,
+				}
+				fileInfo, err := os.Stat(event.Name)
+				if err != nil && fileInfo.IsDir() {
+					dirName = event.Name
+				} else {
+					dirName = filepath.Dir(event.Name)
+				}
+				musicDirName := s.watchMap[dirName]
+				if musicDirName == "" {
+					musicDirName = s.watchMap[filepath.Dir(dirName)]
+				}
+				err = filepath.WalkDir(dirName, func(absPath string, d fs.DirEntry, err error) error {
+					return s.watchCallback(musicDirName, absPath, d, err)
+				})
+				if err != nil {
+					log.Printf("error watching directory tree: %v\n", err)
+				}
+				err = filepath.WalkDir(dirName, func(absPath string, d fs.DirEntry, err error) error {
+					return s.scanCallback(c, musicDirName, absPath, d, err)
+				})
```

Comment:
not sure i completely get what's going on there, why do we need to walk the directory twice when we get a new event?

Reply:
One walk sets the new watch for any created subdirs (since fsnotify is not recursive) and the other walk actually executes the scan.